Engine/src/Engine/ResourceManagement/AdvancedShaderDistributor.cpp

```diff
@@ -0,0 +1,294 @@
+			//with the corresponding data type
+			SourceExtractor* se = ((loadingVertex) ? &vertexExtractor : &fragmentExtractor);
+			switch (pendingMessenger)
+			{
+			case SEMessenger::PENDING_LAYOUT:
+				se->layout.push_back(src[i]);
+				break;
+			case SEMessenger::PENDING_IN_VARIABLES:
+				se->inVariables.push_back(src[i]);
+				break;
+			case SEMessenger::PENDING_OUT_VARIABLES:
+				se->outVariables.push_back(src[i]);
+				break;
+			case SEMessenger::PENDING_UNIFORMS:
+				se->uniforms.push_back(src[i]);
+				break;
+			case SEMessenger::PENDING_MAIN_COMMANDS:
+				se->mainCommands.push_back(src[i]);
+				break;
+			default:
+				//if nothing is pended, don't do anything
+				break;
+			}
+		}
+	}
+
+	void AdvancedShaderDistributor::compileSource(SourceExtractor& to, SourceExtractor& from)
+	{
+		for (int i = 0; i < from.layout.size(); i++)
+		{
+			to.layout.push_back(from.layout[i]);
+		}
+		for (int i = 0; i < from.inVariables.size(); i++)
+		{
+			to.inVariables.push_back(from.inVariables[i]);
+		}
+		for (int i = 0; i < from.outVariables.size(); i++)
+		{
+			to.outVariables.push_back(from.outVariables[i]);
+		}
+		for (int i = 0; i < from.uniforms.size(); i++)
+		{
+			to.uniforms.push_back(from.uniforms[i]);
+		}
+		//pop ending } of main
+		to.mainCommands.pop_back();
+		for (int i = 0; i < from.mainCommands.size(); i++)
+		{
+			to.mainCommands.push_back("\t" + from.mainCommands[i]);
+		}
+		//reinsert }
+		to.mainCommands.push_back("}");
+	}
+
+	void AdvancedShaderDistributor::compileSource(std::string& to, SourceExtractor from)
+	{
+		for (int i = 0; i < from.layout.size(); i++)
+		{
+			to += "layout (location = " + to_string(i) + ") in " + from.layout[i] + "\n";
+		}
+		for (int i = 0; i < from.inVariables.size(); i++)
+		{
+			to += "in " + from.inVariables[i] + "\n";
+		}
+		for (int i = 0; i < from.outVariables.size(); i++)
+		{
+			to += "out " + from.outVariables[i] + "\n";
+		}
+		for (int i = 0; i < from.uniforms.size(); i++)
+		{
+			to += "uniform " + from.uniforms[i] + "\n";
+		}
+		for (int i = 0; i < from.mainCommands.size(); i++)
+		{
+			to += from.mainCommands[i] + ((i < from.mainCommands.size() - 1) ? "\n" : "");
+		}
+	}
+
+	vector<string> AdvancedShaderDistributor::split(string source)
+	{
+		stringstream ss(source);
+		vector<string> list;
+
+		string s;
+		while (getline(ss, s, '\n')) {
+			list.push_back(s);
+		}
+
+		return list;
+	}
+
+	bool AdvancedShaderDistributor::strContains(string parent, string substr)
+	{
+		return parent.find(substr) != string::npos;
+	}
+
+	string AdvancedShaderDistributor::substringAtDataDeclaration(string parent)
+	{
+		string keyword = "";
+		for (int i = 0; i < VARIABLE_TYPES.size(); i++)
+		{
+			if (strContains(parent, VARIABLE_TYPES[i]))
+			{
+				keyword = VARIABLE_TYPES[i];
+				break;
+			}
+		}
+		if (keyword == "")
+		{
+			return keyword;
+		}
+
+		return parent.substr(parent.find(keyword));
```

Comment:
could use a ternary here to be more minimalist, just a suggestion

Reply:
Good work.